lib/projfs.c

```diff
@@ -1451,10 +1505,67 @@ static int check_safe_rel_path(const char *path)
 	return 1;
 }
 
-int projfs_create_proj_dir(struct projfs *fs, const char *path, mode_t mode)
+#define PROJ_XATTR_READ 0x00
+#define PROJ_XATTR_WRITE 0x01
+#define PROJ_XATTR_CREATE 0x02
+
+static char *make_user_xattr_name(const char *segments, unsigned int flags)
+{
+	char *name;
+
+	name = malloc(PROJ_XATTR_PRE_LEN + strlen(segments) + 1);
+	if (name == NULL)
+		return NULL;
+
+	memcpy(name, PROJ_XATTR_PRE_NAME, PROJ_XATTR_PRE_LEN);
+	strcpy(name + PROJ_XATTR_PRE_LEN, segments);
+
+	// permit read of all attributes, but write of non-reserved ones only
+	if (flags == PROJ_XATTR_READ || !xattr_name_is_reserved(name))
+		return name;
```

Comment:
The weirdness has become more clear here; it constructs a name out of pieces, but it also performs access control in the same breath. I think it would be good if we contracted this function's purview to constructing the name only, and then separately did a permissions check on the resulting name/flag combination.

Reply:
I can do that, for sure.  I do want the callers do call into a single function, not two, though.

Reply:
Actually, on a fresh look today I see that I don't have multiple callers of this function anymore, which makes things easy!  I think these revisions should be to your liking, @kivikakk -- let me know!